### PR TITLE
[RFR] Support routing in TabbedForm & TabbedShowLayout

### PR DESCRIPTION
**Pull request by djhi**
This PR adds support for routing driven tab selection both in `TabbedForm` and `TabbedShowLayout`.

- [x] Remove unecessary `exact` from `Resource` to allow urls such as `/posts/10/show/comments`
- [x] Support tab routing by default in `TabbedForm`. Use `path` prop if provided and fallback on tab index if not
- [x] Support tab routing by default in `TabbedShowLayout`. Use `path` prop if provided and fallback on tab index if not
- [x] Tests

Screencast

### CODE REVIEW

**Comment by nicholasio** [comment]
👍

**Comment by fzaninotto** [comment]
I'd prefer if the tabbed layouts supported routing by default. This should work out of the box, without any additional code in the `simple` example.

**Comment by djhi** [comment]
Ok, should we try to infer the pathname for each tab from it's label and allow overriding it with a `pathname` prop ?

**Comment by fzaninotto** [comment]
Why can't we use the tab index in the URL instead?

**Comment by djhi** [comment]
> Why can't we use the tab index in the URL instead?

1. Because indexes in url are ugly
2. Because it's better from a usability perspective to have a text matching the active tab, IMO
3. Because people will ask to customize it

**Comment by djhi** [comment]
I tried adding routing support by default on the TabbedForm. Let me know what you think. I'll tackle the TabbedShowLayout asap

**Comment by christiaanwesterbeek** [comment]
If you use tab names in the URL in a multilingual app with, you'll mix languages. Since resource names probably remain English and the tab names get translated.

**Comment by djhi** [comment]
@fzaninotto I also tried to have the same way to handle tabs in `TabbedForm` and `TabbedShowLayout`, removing the `context` hack in the `Tab` component

File: packages/ra-ui-materialui/src/detail/TabbedShowLayout.js [hunk 1]
```diff
@@ -90,33 +90,50 @@ export class TabbedShowLayout extends Component {
...
-                        (tab, index) =>
-                            tab &&
-                            cloneElement(tab, {
-                                context: 'header',
-                                value: index,
-                            })
-                    )}
+                    {Children.map(children, (tab, index) => {
+                        if (!tab) return null;
+                        const tabPath = `${match.url}${tab.props.path
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/detail/TabbedShowLayout.js]: I'd prefer that you leave that rendering part in `Tab`, in another rendering *context*, as it was done previously. *It's always bad decide too many things for your children* (you can quote me on this).

File: packages/ra-ui-materialui/src/form/TabbedForm.js [hunk 2]
```diff
@@ -111,45 +105,55 @@ export class TabbedForm extends Component {
...
-                                        ) && this.state.value !== index
-                                            ? classes.errorTabButton
-                                            : null
-                                    )}
-                                />
-                            ) : null
-                    )}
+                    {Children.map(children, (tab, index) => {
+                        if (!tab) return null;
+                        const tabPath = `${match.url}${tab.props.path
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/form/TabbedForm.js]: Same remark

File: packages/ra-ui-materialui/src/detail/TabbedShowLayout.js (hunk 1 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/detail/TabbedShowLayout.js]: I'm -1 on this. First it's kind of a hack. Second, our two tabbed components manage their tabs differently. I'd rather have them do the same thing as it is easier to reason about them this way.

Finally, we are still deciding things for our children with this hack imo

File: packages/ra-core/src/Resource.js [hunk 3]
```diff
@@ -115,7 +113,6 @@ export class Resource extends Component {
                 )}
                 {edit && (
                     <Route
-                        exact
```
Comment (fzaninotto) [inline packages/ra-core/src/Resource.js]: doesn't it mean that the route `/posts/12/show` will also render this route?

File: packages/ra-ui-materialui/src/detail/TabbedShowLayout.js (hunk 1 above)
Comment (fzaninotto) [reply packages/ra-ui-materialui/src/detail/TabbedShowLayout.js]: Except it allows to change the way a tab header or content is rendered just by providing a custom `<Tab>` component. With your approach, it's no longer possible. I prefer that we keep the rendering context.

File: packages/ra-core/src/Resource.js (hunk 3 above)
Comment (djhi) [reply packages/ra-core/src/Resource.js]: No because we're in a switch and the show route will take precedence

File: packages/ra-ui-materialui/src/detail/TabbedShowLayout.js (hunk 1 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/detail/TabbedShowLayout.js]: Fair enough. Shouldn't we apply the recipe to the FormTab then ?

However, instead of supplying a custom Tab, shouldn't we suggest a custom Form component altogether when further customization is needed?

File: packages/ra-core/src/Resource.js (hunk 3 above)
Comment (fzaninotto) [reply packages/ra-core/src/Resource.js]: gotcha

File: packages/ra-ui-materialui/src/detail/TabbedShowLayout.js (hunk 1 above)
Comment (fzaninotto) [reply packages/ra-ui-materialui/src/detail/TabbedShowLayout.js]: Both the `Tab` and the `FormTab` should use this double render trick.

Writing a custom Form component requires way more knowledge (redux, redux-form, form name, initial values, etc) than writing a custom tab IMO.

Comment (fzaninotto) [reply packages/ra-ui-materialui/src/detail/TabbedShowLayout.js]: And when I think about it, it makes `SimpleForm` and `TabbedForm` good candidates for a refactoring based on a `FormController` with a render prop.

Comment (djhi) [reply packages/ra-ui-materialui/src/detail/TabbedShowLayout.js]: I was thinking about it too :) However we will probably have to come up with another solution for detecting tabs with errors though

Comment (djhi) [reply packages/ra-ui-materialui/src/detail/TabbedShowLayout.js]: Let's merge this and start a refactoring in another PR ?

Comment (djhi) [reply packages/ra-ui-materialui/src/detail/TabbedShowLayout.js]: Or should I first revert the Tab change and apply the context pattern to the FormTab?

Comment (fzaninotto) [reply packages/ra-ui-materialui/src/detail/TabbedShowLayout.js]: Yes, please readd context, and we’ll refactor the controller part in another PR. 

I’m waiting for this (#1938) PR merge to release 2.1.0.

File: packages/ra-ui-materialui/src/detail/TabbedShowLayout.js [hunk 4]
```diff
@@ -90,33 +88,41 @@ export class TabbedShowLayout extends Component {
             >
                 <Tabs
                     scrollable
-                    value={this.state.value}
+                    value={location.pathname}
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/detail/TabbedShowLayout.js]: :+1: smart way of dealing with tab values. I just think you should add a comment leading to line 97 below because I was scratching my head until uI read the rest of the component source.

File: packages/ra-ui-materialui/src/form/TabbedForm.spec.js [hunk 5]
```diff
@@ -9,28 +9,11 @@ const translate = label => label;
 const muiTheme = { textField: { errorColor: 'red' } };
 
 describe('<TabbedForm />', () => {
-    it('should render tabs', () => {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/form/TabbedForm.spec.js]: why did you remove this one?

Comment (djhi) [reply packages/ra-ui-materialui/src/form/TabbedForm.spec.js]: Because of React Router, it seems that it doesn't render in tests
